src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,241 @@
+                    context.Features.Set<IHttpResponseBodyFeature>(responseBufferingStream);
+                }
+
+                await _next(context);
+
+                if (requestBufferingStream?.HasLogged == false)
+                {
+                    // If the middleware pipeline didn't read until 0 was returned from readasync,
+                    // make sure we log the request body.
+                    requestBufferingStream.LogRequestBody();
+                }
+
+                if (responseBufferingStream == null || responseBufferingStream.FirstWrite == false)
+                {
+                    // No body, write headers here.
+                    LogResponseHeaders(response, options, _logger);
+                }
+
+                if (responseBufferingStream != null)
+                {
+                    var responseBody = responseBufferingStream.GetString(responseBufferingStream.Encoding);
+                    if (!string.IsNullOrEmpty(responseBody))
+                    {
+                        _logger.ResponseBody(responseBody);
+                    }
+                }
+            }
+            finally
+            {
+                responseBufferingStream?.Dispose();
+
+                if (originalBodyFeature != null)
+                {
+                    context.Features.Set(originalBodyFeature);
+                }
+
+                requestBufferingStream?.Dispose();
+
+                if (originalBody != null)
+                {
+                    context.Request.Body = originalBody;
+                }
+            }
+        }
+
+        private static void AddToList(List<KeyValuePair<string, object?>> list, string key, string? value)
+        {
+            list.Add(new KeyValuePair<string, object?>(key, value));
+        }
+
+        public static void LogResponseHeaders(HttpResponse response, HttpLoggingOptions options, ILogger logger)
+        {
+            var list = new List<KeyValuePair<string, object?>>(
+                response.Headers.Count + DefaultResponseFieldsMinusHeaders);
+
+            if (options.LoggingFields.HasFlag(HttpLoggingFields.ResponseStatusCode))
+            {
+                list.Add(new KeyValuePair<string, object?>(nameof(response.StatusCode), response.StatusCode));
+            }
+
+            if (options.LoggingFields.HasFlag(HttpLoggingFields.ResponseHeaders))
+            {
+                FilterHeaders(list, response.Headers, options.ResponseHeaders);
+            }
+
+            var httpResponseLog = new HttpResponseLog(list);
+
+            logger.Log(LogLevel.Information,
+                eventId: LoggerEventIds.ResponseLog,
+                state: httpResponseLog,
+                exception: null,
+                formatter: HttpResponseLog.Callback);
+        }
+
+        internal static void FilterHeaders(List<KeyValuePair<string, object?>> keyValues,
+            IHeaderDictionary headers,
+            ISet<string> allowedHeaders)
```

Comment:
You could use the backing with a concrete type to avoid the virtual dispatch.

Reply:
So you are saying to cast the ISet to a HashSet because I know it's has to be a HashSet? Feels a bit hacky.

Reply:
I meant you could have an internal `HashSet<string>` property and use that here. I guess it depends on how often this gets called, but my guess is this is per request

Reply:
Agree with this. The pattern looks like this:

```C#
public class MyOptions
{
    public ISet<string> SlowPublicThing => FastInternalThing;
    internal HashSet<string> FastInternalThing { get; } = new();
}
```

Then you use `FastInternalThing` internally